### PR TITLE
Bluetooth kernel support

### PR DESCRIPTION
**Pull request by alfert**
Support for bluetooth on kernel level and in providing the bluez utilities via build root. 

To make bluetooth usable from Nerves, we still need an Elixir user land implementation that controls the bluetooth settings via the bluez utilities. But this is not specific to Nerves or the RPI, but depends only on the availability of the bluez stack, i.e. requires a Linux based system. This user land part will appear in another repository (similar to WiFi support). A proof of concept implementation can be found here: https://github.com/alfert/nerves_bluetooth

### CODE REVIEW

**Comment by fhunleth** [comment]
Klaus - this is great work! Bluez is not easy to get integrated.

Given the complexity of some of the components involved here, I'm leaning towards going the slow approach to pull BT support into this system. I'm not aware of any Elixir Bluetooth packages on hex or github, but it would be nice to know near/far we are from people making BLE apps. Is your project going to take you down this route or are you pursuing regular BT?

Also, so that you know, I'd prefer a non-Bluez (BLE-only) approach for Nerves. I don't believe this will happen in the nearterm, so a Bluez/D-bus integration is the best alternative. However, if you ever come across something pointing in a non-Bluez direction, I'd like to know.

**Comment by alfert** [comment]
Frank, thanks! 

I am interested in using BLE to send and receive messages from sensors in an IOT setting. For now, it seems that the bluez way is an rather easy way to come along. I assume that we can hide details about bluez or any other implementation behind a behaviour so that implementations can change without disturbing a client. 

If I understand your non-bluez approach correct, it would mean that we need to port some parts of bluez for loading firmware into the controllers (like the patches I need to add) and providing access to the data sent and received similar to a network device or a tty. Does that mean that we need new  / other kernel drivers as well? That would make difference.

File: nerves_defconfig [hunk 1]
```diff
@@ -49,10 +50,18 @@ BR2_PACKAGE_FWUP=y
...
+BR2_PACKAGE_SETSERIAL=y
 BR2_PACKAGE_ERLANG_SMP=y
 BR2_PACKAGE_LIBMNL=y
-BR2_PACKAGE_NCURSES=y
+BR2_PACKAGE_BLUEZ5_UTILS=y
+BR2_PACKAGE_BLUEZ5_UTILS_OBEX=y
+BR2_PACKAGE_BLUEZ5_UTILS_CLIENT=y
+BR2_PACKAGE_BLUEZ5_UTILS_GATTTOOL=y
+BR2_PACKAGE_BLUEZ5_UTILS_EXPERIMENTAL=y
+BR2_PACKAGE_DNSMASQ=y
```
Comment (fhunleth) [inline nerves_defconfig]: dnsmasq and setserial aren't needed, are they?

File: linux-4.4.defconfig [hunk 2]
```diff
@@ -76,6 +76,19 @@ CONFIG_SYN_COOKIES=y
...
 # CONFIG_INET_DIAG is not set
 # CONFIG_IPV6 is not set
+CONFIG_BT=m
+CONFIG_BT_RFCOMM=m
+CONFIG_BT_RFCOMM_TTY=y
+CONFIG_BT_BNEP=m
+CONFIG_BT_BNEP_MC_FILTER=y
+CONFIG_BT_BNEP_PROTO_FILTER=y
+CONFIG_BT_HIDP=m
+CONFIG_BT_HCIBTUSB=m
```
Comment (fhunleth) [inline linux-4.4.defconfig]: I think that a few of these modules aren't needed. BT_HCIBTUSB doesn't sound like it should be used. Could you run an `lsmod` to see what's actually loaded?

**Comment by fhunleth** [comment]
I know of two non-bluez ways. One is to interface directly to the UART going to the BT module. There are a couple BT stacks that do this. The second route would be to use the Linux kernel's BT stack, but implement the userland parts in Elixir. This second route uses most of your patch and would be the way that I'd prefer if an Elixir BLE library were available that supported it. Both routes don't require additional kernel drivers.

Please keep going down the bluez path, though. I agree with your comment that it should be possible to create a BLE API such that you could switch out the low level implementation.

File: linux-4.4.defconfig (hunk 2 above)
Comment (alfert) [inline linux-4.4.defconfig]: `lsmod` listed the following modules: 
* `hci_uart`
* `btbcm`
* `bnep`
* `bluetooth`

I enabled nearly all BT relevant configurations within the kernel during my debugging sessions. But I never reduced them afterwards to the relevant minimal set.

File: nerves_defconfig (hunk 1 above)
Comment (alfert) [inline nerves_defconfig]: Yes, you are right. At least, I see no reason for them and I did not select them consciously.

A short review of the help texts within `make menuconfig` reveals that
`dnsmasq` is not needed - don't know where it comes from - a dhcp server makes no sense. But `setserial` is required for `BR2_USE_MMU` and I assume that we need a memory management unit for a proper Linux system.

Comment (fhunleth) [inline nerves_defconfig]: Regarding `setserial`, the `BR2_USE_MMU` indicates that `setserial` requires an MMU to work (the Buildroot Config.in file says that it calls `fork()`, and `fork()` doesn't work on the MMU-less ucLinux. `setserial` is a utility program for configuring serial ports. It shouldn't be needed unless there's a shell script somewhere that calls into it. My suspicion is that `setserial` was enabled to help debug a serial port issue and is no longer needed.

Comment (alfert) [inline nerves_defconfig]: Ah, yes, that makes sense: The dependency is the other way round. I will remove `setserial`, it is not needed for this case.

**Comment by fhunleth** [comment]
Closing since this has gotten stale and we're focusing on https://github.com/verypossible/harald. While not as full featured, `harald` integrates much more simply.

**Pull request closed without merging**